### PR TITLE
[plugin] Add Plugin API to communicate between Theia and plugins

### PR DESCRIPTION
**Pull request by svor**
Signed-off-by: Valeriy Svydenko <vsvydenk@redhat.com>

Add Plugin API to communicate between Theia and plugins via `MessageReader` and `MessageWriter`.

### CODE REVIEW

File: packages/plugin-ext/src/common/connection.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { AbstractMessageReader, DataCallback } from 'vscode-jsonrpc/lib/messageReader';
+import { AbstractMessageWriter } from 'vscode-jsonrpc/lib/messageWriter';
+import { ConnectionMain, ConnectionExt } from '../api/plugin-api';
+import { Disposable } from './disposable-util';
+
+export class PluginConnection implements Disposable {
```
Comment (benoitf) [inline packages/plugin-ext/src/common/connection.ts]: could you describe quickly the role of this class in comments ?

File: packages/plugin-ext/src/common/connection.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        this.writer = pluginMessageWriter;
+        this._dispose = _dispose;
+    }
+
+    dispose(): void {
+        this._dispose();
+    }
+}
+
+export class PluginMessageWriter extends AbstractMessageWriter {
```
Comment (benoitf) [inline packages/plugin-ext/src/common/connection.ts]: could we move out these two inner classes(PluginMessageWriter PluginMessageReader ) and  to separate files  as they're exported?

File: packages/plugin-ext/src/main/browser/connection-main.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { MAIN_RPC_CONTEXT, ConnectionMain, ConnectionExt } from '../../api/plugin-api';
+import { RPCProtocol } from '../../api/rpc-protocol';
+import { PluginConnection, PluginMessageReader, PluginMessageWriter } from '../../common/connection';
+
+export class ConnectionMainImpl implements ConnectionMain {
```
Comment (benoitf) [inline packages/plugin-ext/src/main/browser/connection-main.ts]: could you please add a brief comment

File: packages/plugin-ext/src/common/connection.ts (hunk 2 above)
Comment (svor) [reply packages/plugin-ext/src/common/connection.ts]: divided

File: packages/plugin-ext/src/main/browser/connection-main.ts (hunk 3 above)
Comment (svor) [reply packages/plugin-ext/src/main/browser/connection-main.ts]: commented

File: packages/plugin-ext/src/common/connection.ts (hunk 1 above)
Comment (svor) [reply packages/plugin-ext/src/common/connection.ts]: commented

**Comment by svor** [comment]
cc @tolusha @tsmaeder

File: packages/plugin-ext/src/common/connection.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { Disposable } from './disposable-util';
+import { PluginMessageReader } from './plugin-message-reader';
+import { PluginMessageWriter } from './plugin-message-writer';
+
+/**
+ * The container for message reader and writer which can be used to create connection between plugins and main said.
```
Comment (benoitf) [inline packages/plugin-ext/src/common/connection.ts]: I suppose it's a typo on `said`?

File: packages/plugin-ext/src/common/connection.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    reader: PluginMessageReader;
+    writer: PluginMessageWriter;
+    _dispose: () => void;
+
+    constructor(protected readonly pluginMessageReader: PluginMessageReader,
+        protected readonly pluginMessageWriter: PluginMessageWriter,
+        _dispose: () => void) {
+        this.reader = pluginMessageReader;
+        this.writer = pluginMessageWriter;
+        this._dispose = _dispose;
```
Comment (benoitf) [inline packages/plugin-ext/src/common/connection.ts]: please note that _ is forbidden in theia guidelines
https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names

File: packages/plugin-ext/src/common/connection.ts (hunk 4 above)
Comment (svor) [reply packages/plugin-ext/src/common/connection.ts]: it is) fixed

File: packages/plugin-ext/src/common/connection.ts (hunk 5 above)
Comment (svor) [reply packages/plugin-ext/src/common/connection.ts]: Thanks! Fixed

**Pull request merged**